src/enzyme_ad/jax/Passes/AffineToStableHLORaising.cpp

```diff
@@ -547,24 +547,89 @@ bool isSafeToSpeculativelyExecuteAtScope(Operation *scope, Operation *op) {
   return inBounds;
 }
 
+struct ParallelContext {
+  SmallVector<InductionVariableRange, 8> ranges;
+  SmallVector<Value, 8> ivs;
+
+  RankedTensorType getTensorType(Type elTy) {
+    SmallVector<int64_t> shape = llvm::map_to_vector(
+        ranges, [&](auto range) { return range.getNumIters(); });
+    return RankedTensorType::get(shape, elTy);
+  }
+
+  struct Broadcast {
+    Value v;
+    affine::AffineValueMap avm;
+  };
+
+  std::optional<Broadcast> getBroadcast(OpBuilder &b,
+                                        affine::AffineValueMap avm, Value v) {
+    auto CTT = dyn_cast<RankedTensorType>(v.getType());
+    if (!CTT)
+      return std::nullopt;
+    auto TT = getTensorType(CTT.getElementType());
+    assert(CTT.getElementType() == TT.getElementType());
+    if (CTT.getShape() == TT.getShape())
+      return Broadcast{v, avm};
+    if (llvm::any_of(llvm::zip(CTT.getShape(), TT.getShape()),
+                     [](auto p) { return std::get<0>(p) != std::get<1>(p); }))
+      return std::nullopt;
+    if (CTT.getRank() > TT.getRank())
+      return std::nullopt;
+
+    // TODO I haven't thought through how to broadcast non-scalars to the
+    // shape we need.
+    if (CTT.getRank() != 0)
+      return std::nullopt;
+    SmallVector<int64_t> dimsToBroadcast;
+    auto bc = b.create<stablehlo::BroadcastInDimOp>(v.getLoc(), TT, v,
+                                                    dimsToBroadcast);
+
+    AffineMap newMap = AffineMap::getMultiDimIdentityMap(
+        TT.getRank() - CTT.getRank(), b.getContext());
+
+    return Broadcast{bc, affine::AffineValueMap(newMap, ivs)};
```

Comment:
@Pangoraw I need your input here. We use this function to "expand" any tensor that is incoming as an initial value for an affine.for iter_arg to the iteration space of the parallel we are in. I currently only handle the case where it already exactly matches what we need, or it is a scalar. I don't know what else we can encounter wrt to the incoming shapes and maps in here. 

Would you be able to help with this implementation?

(see the test cases for reference for cases we can handle)

Reply:
That looks good, I think we can start with this case of only supporting constants and see where that gets us. If that's not sufficient we can still broadcast the multiple init values with the right broadcast dims.